### ISSUE TITLE
Qubes widgets should work properly in Xwayland

### ISSUE THREAD

**Issue by tasket** [QubesOS/qubes-issues#8410]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### Qubes OS release

4.2rc1, updated with current-testing including:

kde-settings-qubes-5.12.3-8
plasma-breeze-qubes-5.12.3-8

### Brief summary
After installing and switching to KDE in dom0, none of the Qubes widgets such as _qui-domains_ and _qui-devices_ are visible/usable.

Note that KDE now defaults to Wayland and this issue does not occur when selecting an X11 session.

### Steps to reproduce
Install KDE: `sudo qubes-dom0-update kde-settings-qubes`

Also make the changes for lightdm->sddm.

On edit:  Also select 'Wayland' session when logging in.

### Expected behavior
Qubes widgets appear when logging in to desktop.

### Actual behavior
Somewhat normal KDE startup, except no Qubes widgets (and the App Launcher appears in the center of the screen, not near the widget).  They are not listed under the hidden menu, and configuring systray doesn't list any Qubes widgets.  The standard KDE widgets are present.

### Related
Issues #8326 #8159 #8187 #3366

**Comment by tasket** [comment]
If I try to run `qui-devices` directly in the shell I get:
```
Gtk-CRITICAL **: gtk_widget_get_scale_factor: assertion 'GTK_IS_WIDGET (widget)' failed
```

...and then two theme parsing errors for 'gtk.css' saying '-gtk-icon-size is not a valid property name'

**Comment by tasket** [comment]
Switching the desktop session from Wayland to X11 appears to get everything working again.  It appears recent versions of KDE default to Wayland.

The KDE doc probably should be updated to instruct the user to select X11.

**Comment by tasket** [comment]
In that case, this might be a duplicate of a generic Wayland issue if one already exists.

**Comment by DemiMarie** [comment]
> In that case, this might be a duplicate of a generic Wayland issue if one already exists.

Not a duplicate, but possibly a subtask.

**Comment by marmarta** [comment]
It looks like a Wayland problem. It does not support tray status icons.

**Comment by marmarta** [comment]
This is blocked by all gtk statusnotifier icons that I know of (libappindicator and libayatana-appindicator) not supporting tooltips on the icon.

**Comment by DemiMarie** [comment]
@marmarta Alternative is to use underlying D-Bus API directly; I’m willing to help with that.  The C libraries just use D-Bus under the hood.

**Comment by DemiMarie** [comment]
There are a few problems here:

1. The menus under KDE try to use Wayland, rather than X11, but this fails rather miserably.  Workaround: `WAYLAND_DISPLAY=` in the environment.
2. KDE + Wayland doesn’t support X11 tray icons natively.  It tries to translate them to StatusNotifierItem, but this translation is buggy and does not work reliably.
3. Ayatana indicators are not packaged in Fedora.
4. Ayatana indicators do not support tooltips, but there is a PR (AyatanaIndicators/libayatana-appindicator#17) that adds that support.
5. AyatanaIndicators/libayatana-appindicator#17 PR is known to cause segfaults under GNOME when [Ubuntu’s GNOME tray icon extension](https://github.com/ubuntu/gnome-shell-extension-appindicator), but seems to work under KDE and XFCE.  This is almost certainly a memory unsafety problem in C.

Personally, my preference remains writing the whole thing in Python, mostly because that means that both @marmarta and I can actually understand what is happening, and that problems show up as Python exceptions and not random segfaults.

**Comment by marmarta** [comment]
well, as usual, I don't think writing and supporting a whole library up from scratch is a feasible solution. For me it just sounds like KDE+Wayland is not really working yet.

**Comment by DemiMarie** [comment]
@marmarta want me to report a bug in KDE?

**Comment by marmarta** [comment]
> @marmarta want me to report a bug in KDE?

You're the person testing it - and you know that all I know about it is what we tested together on a call. Do what you think is right.

**Comment by DemiMarie** [comment]
Tray icons not working is definitely a KDE bug: <https://bugs.kde.org/show_bug.cgi?id=468085>.  There are two options I can think of, and both should avoid needing to rewrite all of the widgets:

1. Switch to Ayatana indicators and use the out-of-tree patch <https://github.com/AyatanaIndicators/libayatana-appindicator/pull/17>.
2. Fix the bug in KDE and run the tray icon under Xwayland.

**Comment by marmarek** [comment]
The current plan is for @DemiMarie to find the best solution _and_ prepare a PoC that shows it actually works, and how to use it in Python.

Pasting a message from @marmarta: basic requirements for the POC are:

- python bindings for Gtk3, included in fedora that we have in dom0
- tooltips for the tray icon
- working icon that changes when directed to
- dynamic menu with items that contain any arbitrary Gtk.Widget
- the menu hides and shows in the same way as in X - hides when you click anywhere outside it
- the submenus show where there is actual visible space, don't hide if there is not enough space on the right/left
- CSS styling support to the extent Gtk3 normally covers it

**Comment by DemiMarie** [comment]
> The current plan is for @DemiMarie to find the best solution _and_ prepare a PoC that shows it actually works, and how to use it in Python.

For the widgets, I will create a full-screen transparent window.  This will not affect behavior under X11 and will fix behavior under Wayland.

The app menu will use wlr-layer-shell.

**Comment by DemiMarie** [comment]
@marmarek: Was dom0 widgets not closing properly the _only_ widget-related bug that you found with the KDE Wayland session?  When I did my testing, lots of other stuff did not work, but my testing was back in March and used an older version of Plasma.  If things work with the newer Plasma in Fedora 41, this means that the [xembedsniproxy bug](https://bugs.kde.org/show_bug.cgi?id=468085) can be closed.

**Comment by marmarek** [comment]
On Sat, Nov 30, 2024 at 01:50:45PM -0800, Demi Marie Obenour wrote:
> @marmarek: Was dom0 widgets not closing properly the _only_ widget-related bug that you found with the KDE Wayland session?  When I did my testing, lots of other stuff did not work, but my testing was back in March and used an older version of Plasma.  If things work with the newer Plasma in Fedora 41, this means that the [xembedsniproxy bug](https://bugs.kde.org/show_bug.cgi?id=468085) can be closed.

That's the only issue with dom0 widgets I found, but I haven't tested it very thoroughly.

The widgets from VMs (like nm-applet) have more issues, for example
nm-applet shows as solid red instead of its icon (but its menu seems to
work just fine).

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab

**Comment by marmarek** [comment]
This is fixed in https://github.com/QubesOS/qubes-desktop-linux-manager/pull/234
